existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,264 @@
+The only solution here is to add constraints and `forall.` quantifier to the log function signature:
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now we put all responsibility for deducing types on the shoulders of the compiler, and it will deduce them the way it wants.
+
+#### Hiding type variables
+
+```Haskell
+data MyData a = forall b. MkMyData b a
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+```
+
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of the expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+The compiler will deduce the type of the variable itself, and you can't specify it.
+
+```haskell
+func :: MyData Integer -> IO ()
+func (MkMyData (x :: Double) y) = print x -- compiler error - can't match type of x with Double
+```
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+
+This might be useful when building a multi-typed list:
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+This is a heterogeneous list that contains values of different types.
+You can't do that with an universally quantified list because every element in that list would need to have the same type.
+But, there are not a lot of things you can do with elements of this list.
+Since the type of elements is unknown, you can't address it, nor can you pass it to functions that expect values of concrete type.
+However, adding constraints to the `a` variable may improve the situation:
+
+```haskell
+data Elem = forall a. (Show a) => Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+
+printElem :: Elem -> IO ()
+printElem (Elem x) = print x
+
+main :: IO ()
+main = fmap printElem multitypedList
+
+λ> "a"
+λ> 1
+λ> Just 5
+```
+
+Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.
+
+```haskell
+showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
+showFunc (x:xs) => (show x) ++ (showFunc xs)
+```
+
+We hide exception type in the same way in the well-known `SomeException` wrapper:
+
+```haskell
+data SomeException = forall a. Exception a => SomeException a
+```
+
+It allows you to catch exceptions of any type.
+As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be catch, wrapped in `SomeException` datatype.
+After that, you can use `fromException` function, to check, that you caught something you wanted to catch.
+
+
+## Summary
```

Comment:
Good post. Not much left to fix.
Pay special attention to Diogo's feedback, dig the history of GHC trac ticket #26 and explain why exists was replaced by just forall. Do some minor tweaks as per my review too.

Amazing job, @PruStephan !